### PR TITLE
New resource: `azurerm_key_vault_managed_hardware_security_module_key_rotation_policy`

### PR DESCRIPTION
**Pull request by wuxu92**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description

Add key rotation policy support for Managed HSM Keys. Since this based on a Data plane API, I created a New resource for it instead of integrating into the Managed HSM Key resource. Based on my test, notification is not supported in MHSM Key. There is only one rotate policy for each key and there is no separation resource/id for it, so the new resource just resuse the managed hsm key id as its resource id.

https://learn.microsoft.com/en-us/azure/key-vault/managed-hsm/key-rotation

Given that creating a Managed HSM resource is a heavy operation, I created only one AccTest with both basic and update included. The creation/update of the rotation policy is fast.

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [ ] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

```
--- PASS: TestAccMHSMKeyRotationPolicy_all (2749.04s)
PASS
```

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_key_vault_managed_hardware_security_module_key_rotation_policy` - New resource for Managed HSM Key rotation policy

This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change

## Related Issue(s)

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: website/docs/r/key_vault_managed_hardware_security_module_key_rotation_policy.html.markdown [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  expire_after       = "P60D"
+  time_before_expiry = "P30D"
+}
+```
+
+## Arguments Reference
+
+The following arguments are supported:
+
+* `expire_after` - (Required) Specify the expiration time of the key.
```
Comment (katbyte) [inline website/docs/r/key_vault_managed_hardware_security_module_key_rotation_policy.html.markdown:35]: is this a date? a duration? what format?
```suggestion
* `expire_after_date` - (Required) Specify the expiration time of the key.
```
```suggestion
* `expire_after_duration` - (Required) Specify the expiration time of the key.
```

File: website/docs/r/key_vault_managed_hardware_security_module_key_rotation_policy.html.markdown [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+* `managed_hsm_key_id` - (Required) The ID of the Managed HSM Key. Changing this forces a new Key Vault to be created.
+
+---
+
```
Comment (katbyte) [inline website/docs/r/key_vault_managed_hardware_security_module_key_rotation_policy.html.markdown:40]: why is this sperator here?

File: website/docs/r/key_vault_managed_hardware_security_module_key_rotation_policy.html.markdown (hunk 1 above)
Comment (katbyte) [reply website/docs/r/key_vault_managed_hardware_security_module_key_rotation_policy.html.markdown:35]: please detail what this actually means, is it related to the lifecycle actions/trigger not the key?

File: internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource.go [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+				return fmt.Errorf("decoding: %+v", err)
+			}
+
+			keyID, err := parse.ManagedHSMDataPlaneVersionlessKeyID(config.ManagedHSMKeyID, domainSuffix)
+			if err != nil {
+				return fmt.Errorf("parsing Managed HSM Key ID: %+v", err)
+			}
+
+			_, err = client.GetKey(ctx, keyID.BaseUri(), keyID.KeyName, "")
+			if err != nil {
```
Comment (katbyte) [inline internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource.go:112]: these lines can be merged

File: internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource.go [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+		expiryTime = utils.String(policy.ExipreAfter)
+	}
+
+	lifetimeActions := make([]keyvault.LifetimeActions, 0)
+
+	lifetimeActionRotate := keyvault.LifetimeActions{
+		Action: &keyvault.LifetimeActionsType{
+			Type: keyvault.ActionTypeRotate,
+		},
+		Trigger: &keyvault.LifetimeActionsTrigger{},
```
Comment (katbyte) [inline internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource.go:248]: are there other actions possible here? what about for trigger?

File: internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource.go [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+		Attributes: &keyvault.KeyRotationPolicyAttributes{
+			ExpiryTime: expiryTime,
+		},
+	}
+}
+
+func flattenKeyRotationPolicy(p keyvault.KeyRotationPolicy) MHSMKeyRotationPolicyResourceSchema {
+	var res MHSMKeyRotationPolicyResourceSchema
+	if p.Attributes != nil {
+		res.ExipreAfter = pointer.From(p.Attributes.ExpiryTime)
```
Comment (katbyte) [inline internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource.go:270]: what does this expire? what happens ?

File: website/docs/r/key_vault_managed_hardware_security_module_key_rotation_policy.html.markdown [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+* `time_after_creation` - (Optional) Specify the rotation time after creation in the format of ISO86061. Exactly one of `time_after_creation` or `time_before_expiry` should be specified.
```
Comment (katbyte) [inline website/docs/r/key_vault_managed_hardware_security_module_key_rotation_policy.html.markdown:41]: this is not clear, it should be explicit in what it is doing when
```suggestion
* `time_after_creation_to_rotate` - (Optional) Specify the rotation time after creation in the format of ISO86061. Exactly one of `time_after_creation` or `time_before_expiry` should be specified.
```

also is this ISO 8601 duration not ISO86061?

File: website/docs/r/key_vault_managed_hardware_security_module_key_rotation_policy.html.markdown [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+* `time_before_expiry` - (Optional) Specify the rotation time before expiration in the format of ISO86061. Exactly one of `time_after_creation` or `time_before_expiry` should be specified.
```
Comment (katbyte) [inline website/docs/r/key_vault_managed_hardware_security_module_key_rotation_policy.html.markdown:43]: same here
```suggestion
* `time_before_expiry_to_rotate` - (Optional) Specify the rotation time before expiration in the format of ISO86061. Exactly one of `time_after_creation` or `time_before_expiry` should be specified.
```

**Review by katbyte — CHANGES_REQUESTED** [review]

File: internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource.go [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+			return nil
+		},
+	}
+}
+
+func expandKeyRotationPolicy(policy MHSMKeyRotationPolicyResourceSchema) keyvault.KeyRotationPolicy {
+
+	var expiryTime *string // = nil // needs to be set to nil if not set
+	if policy.ExipreAfter != "" {
+		expiryTime = utils.String(policy.ExipreAfter)
```
Comment (jackofallops) [inline internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource.go:239]: Can we not use the `utils` package here as we're deprecating these functions, and instead use the `pointer` package as you have done in the flatten below?
```suggestion
		expiryTime = pointer.To(policy.ExipreAfter)
```

File: internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource_test.go [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+	"github.com/hashicorp/terraform-provider-azurerm/internal/clients"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/managedhsm/parse"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+	"github.com/hashicorp/terraform-provider-azurerm/utils"
+)
+
+type MHSMKeyRotationPolicyTestResource struct{}
+
+func testAccMHSMKeyRotationPolicy_all(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_key_vault_managed_hardware_security_module_key", "test")
```
Comment (jackofallops) [inline internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource_test.go:23]: This is the wrong resource here?
```suggestion
	data := acceptance.BuildTestData(t, "azurerm_key_vault_managed_hardware_security_module_key_rotation_policy", "test")
```

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @wuxu92 - Just a couple of additional comments to @katbyte 's review to take a look at.

File: internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource.go [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+			if !ok {
+				return fmt.Errorf("could not determine Managed HSM domain suffix for environment %q", metadata.Client.Account.Environment.Name)
+			}
+
+			id, err := parse.ManagedHSMDataPlaneVersionlessKeyID(metadata.ResourceData.Id(), domainSuffix)
+			if err != nil {
+				return err
+			}
+
+			// settign a blank policy to delete the existing policy
```
Comment (jackofallops) [inline internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource.go:221]:
```suggestion
			// setting a blank policy to delete the existing policy
```

File: internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource.go [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+
+var _ sdk.ResourceWithUpdate = KeyVaultMHSMKeyRotationPolicyResource{}
+
+func (r KeyVaultMHSMKeyRotationPolicyResource) ModelObject() interface{} {
+	return &MHSMKeyRotationPolicyResourceSchema{}
+}
+
+type MHSMKeyRotationPolicyResourceSchema struct {
+	ManagedHSMKeyID   string `tfschema:"managed_hsm_key_id"`
+	ExipreAfter       string `tfschema:"expire_after"`
```
Comment (jackofallops) [inline internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource.go:32]: Typo here
```suggestion
	ExpireAfter       string `tfschema:"expire_after"`
```

File: internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource_test.go [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+
+resource "azurerm_key_vault_managed_hardware_security_module_key_rotation_policy" "test" {
+  managed_hsm_key_id  = azurerm_key_vault_managed_hardware_security_module_key.test.id
+  expire_after        = "P60D"
+  time_after_creation = "P28D"
+}
+
+
+
+
```
Comment (jackofallops) [inline internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource_test.go:105]: unnecessary whitespace here?
```suggestion
```

File: internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource_test.go [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+        "keyAgreement",
+        "keyCertSign",
+        "keyEncipherment",
+      ]
+      subject            = "CN=hello-world"
+      validity_in_months = 12
+    }
+  }
+}
+resource "azurerm_key_vault_managed_hardware_security_module" "test" {
```
Comment (jackofallops) [inline internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource_test.go:196]:
```suggestion
}

resource "azurerm_key_vault_managed_hardware_security_module" "test" {
```

File: internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource_test.go [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+
+%s
+
+resource "azurerm_key_vault_managed_hardware_security_module_key_rotation_policy" "test" {
+  managed_hsm_key_id = azurerm_key_vault_managed_hardware_security_module_key.test.id
+  time_before_expiry = "P30D"
+  expire_after       = "P60D"
+}
+
+
```
Comment (jackofallops) [inline internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource_test.go:85]: unnecessary whitespace here?
```suggestion
```

File: internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource_test.go [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+      "Get",
+      "Purge",
+      "Update"
+    ]
+  }
+  tags = {
+    environment = "Production"
+  }
+}
+resource "azurerm_key_vault_certificate" "cert" {
```
Comment (jackofallops) [inline internal/services/managedhsm/key_vault_managed_hardware_security_module_key_rotation_policy_resource_test.go:156]:
```suggestion
}

resource "azurerm_key_vault_certificate" "cert" {
```

**Review by jackofallops — CHANGES_REQUESTED** [review]
Thanks for the updates @wuxu92 - Spotted a couple typos and whitespace items if you can take a look, then I think we should be good to run the test.

**Review by katbyte — APPROVED** [review]
Thanks @wuxu92  - LGTM 🛋️

**Pull request merged**